core/state_transition.go

```diff
@@ -504,20 +504,30 @@ func (st *StateTransition) TransitionDb(refunds bool, gasBailout bool) (*evmtype
 	} else {
 		ret, st.gasRemaining, vmerr = st.evm.Call(sender, st.to(), st.data, st.gasRemaining, st.value, bailout)
 	}
-	gasUsed := st.gasUsed()
-	if gasUsed < floorGas7623 && rules.IsPrague {
-		gasUsed = floorGas7623
-		st.gasRemaining = st.initialGas - gasUsed
-	}
+
+	var refundQuotient uint64
 	if refunds && !gasBailout {
 		if rules.IsLondon {
 			// After EIP-3529: refunds are capped to gasUsed / 5
-			st.refundGas(params.RefundQuotientEIP3529)
+			refundQuotient = params.RefundQuotientEIP3529
 		} else {
 			// Before EIP-3529: refunds were capped to gasUsed / 2
-			st.refundGas(params.RefundQuotient)
+			refundQuotient = params.RefundQuotient
+		}
+		gasUsed := st.gasUsed()
+		// Apply refund counter, capped to half of the used gas.
+		refund := gasUsed / refundQuotient
```

Comment:
with golang 1.21 it can be simplified to
`
refund := min(st.gasUsed()/refundQuotient, st.state.GetRefund())
`